### PR TITLE
Switch CatalogEntity and CatalogCategory abstract classes

### PR DESCRIPTION
**Pull request by Nokel81**
Signed-off-by: Sebastian Malton <sebastian@malton.name>

### CODE REVIEW

File: src/common/catalog-entities/kubernetes-cluster.ts [hunk 1]
```diff
@@ -81,19 +75,19 @@ export class KubernetesCluster implements CatalogEntity {
...
     icon: require(`!!raw-loader!./icons/kubernetes.svg`).default // eslint-disable-line
   };
-  public spec = {
+  public spec: CatalogCategorySpec = {
     group: "entity.k8slens.dev",
     versions: [
       {
         name: "v1alpha1",
-        entityClass: KubernetesCluster
+        EntityClass: KubernetesCluster
```
Comment (jakolehm) [inline src/common/catalog-entities/kubernetes-cluster.ts]:
```suggestion
        entityClass: KubernetesCluster
```

File: src/common/catalog-entities/web-link.ts [hunk 2]
```diff
@@ -56,7 +50,7 @@ export class WebLinkCategory implements CatalogCategory {
     versions: [
       {
         name: "v1alpha1",
-        entityClass: WebLink
+        EntityClass: WebLink
```
Comment (jakolehm) [inline src/common/catalog-entities/web-link.ts]:
```suggestion
        entityClass: WebLink
```

File: src/common/catalog/catalog-entity.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  (new (data: CatalogEntityData<
+    ExtractEntityMetadataType<Entity>,
+    ExtractEntityStatusType<Entity>,
+    ExtractEntitySpecType<Entity>
+  >) => Entity)
+);
+
+export interface CatalogCategoryVersion<Entity extends CatalogEntity> {
+  name: string;
+  EntityClass: CatalogEntityConstructor<Entity>;
```
Comment (jakolehm) [inline src/common/catalog/catalog-entity.ts]:
```suggestion
  entityClass: CatalogEntityConstructor<Entity>;
```

File: src/common/catalog/catalog-entity.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+export interface CatalogCategorySpec {
+  group: string;
+  versions: CatalogCategoryVersion<CatalogEntity>[];
+  names: {
+    kind: string;
+  };
+}
+
+export abstract class CatalogCategory extends EventEmitter {
```
Comment (jakolehm) [inline src/common/catalog/catalog-entity.ts:29]: Maybe split out `CatalogCategory` things to a separate while now that we are moving these anyway.

**Review by jakolehm — APPROVED** [review]
LGTM

**Review by jim-docker — APPROVED** [review]
Barely related to this PR, I find it confusing that we have two `CatalogEntityRegsitry`'s. It would be nice if we could rename the `common` one, but I'm not sure what to. `CatalogEntityMainRegsitry`?

**Comment by Nokel81** [comment]
@jim-docker yeah... I think the one in `extensions` is supposed to be "facade" one.

**Comment by jim-docker** [comment]
> @jim-docker yeah... I think the one in `extensions` is supposed to be "facade" one.

Or "shadow", I thought of that too, but that one is part of the API right?  So we probably want to keep that name as is.

**Pull request merged**